### ISSUE TITLE
Completed To-Dos disappear if you sync after viewing them

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#9806]
Completed To-Dos behave in a way that appears to be inconsistent and buggy although it's a sort-of-expected side-effect of the way that we don't load completed To-Dos at all until the user clicks on the "Complete" tab in the To-Dos column.

When the user does that, the completed To-Dos are read from the database and they appear after that data is synced to the website. However if the user clicks the website's sync button while they are viewing the "Complete" tab, the completed To-Dos vanish (I assume this is because the full-account sync causes the To-Dos data to be replaced with all active To-Dos and no completed ones). If the user moves to the Active or Scheduled tab and then back to Complete, the completed To-Dos do NOT reappear. The user has to reload the website to be able to view completed To-Dos again.

The reload should not be necessary.

Every time the user goes to the "Complete" tab, their completed To-Dos should always be reloaded from the database.

As well as that, ideally if the user clicks the sync button while the "Complete" tab is active, then the completed To-Dos should be synced along with the uncompleted ones. However it's possible that that might be too difficult to do - I haven't looked into it whether the sync action can get information about whether the "Complete" tab is visible.

----

_It's possible that there are also other bugs as indicated in the messages below from Kella (b50b5e4d-0134-4e3b-9627-565840558f59) in the Report a Bug guild. However we should fix the sync behaviour first and then investigate further to see if more fixes are needed - if so, we can log a new issue once this one is closed._

_Kella: "On the tasks page, in the To-Do's column, when I select the "Complete" tab, the list of tasks goes blank. The to-do's I've completed recently show up in the Data Display tool just fine but after refreshing the page, the Completed list of to-do's is still empty."_

_Alys: "Is it possible it's just taking a while for the completed To-Dos to appear? When I view them, they do appear but only after a couple of seconds. Can you try staying in the completed list for ten seconds to see if that makes a difference?"_

_Kella: "I tried waiting 10 seconds for my completed to-do's to show up and nothing. Since then, I checked another time and only the to-do I completed that day showed up. Then I tried syncing the page, which caused that completed to-do to disappear again. Then I refreshed the webpage, and the to-do's all appeared. This morning, went back to look and all the completed to-do's were gone again. Refreshed the website again and they came back."_

**Comment by Alys** [comment]
I'm reopening this because there's an intermittent problem in staging:

1. Go to the Tasks page and reload it.
2. Click on the To-Dos "Complete" tab. Observe that completed To-Dos appear.
3. Click the sync button. Observe the one of these will happen (seems to be random):
   a. All completed To-Dos disappear (the most common result from my testing)
   b. All completed To-Dos reappear briefly but then disappear again
   c. All completed To-Dos reappear correctly and remain visible

The desired result is (c).

If the completed To-Dos disappear, then if you go to the "Active" tab and then back to "Completed", the completed To-Dos are still invisible. The desired result is that they should reappear.

**Comment by Mahendran-Nadesan** [comment]
I'd like to work on this.

**Comment by paglias** [comment]
It's yours :)

**Comment by beffymaroo** [comment]
@Mahendran-Nadesan Are you still working on this issue? Please let us know. Thank you!

**Comment by Mahendran-Nadesan** [comment]
Hi, please see my comment here: https://github.com/HabitRPG/habitica/issues/9904#issuecomment-382712136

**Comment by beffymaroo** [comment]
@Mahendran-Nadesan Thank you for the updates!

**Comment by beffymaroo** [comment]
@Mahendran-Nadesan Checking in here as well- are you still working on this issue?

**Comment by Mahendran-Nadesan** [comment]
Hi, I think it's best I relinquish this task so someone else can do it.
